### PR TITLE
terra: migrate to cosmwasm 1.1.0

### PR DESCRIPTION
**Pull request by kcsongor**
Terra classic upgraded to CosmWasm 1.0+ (https://classic-agora.terra.money/t/wasmvm-v0-16-6-v1-2-0/50070/73), which introduced a number of breaking changes across the whole ecosystem, ranging from tooling to smart contracts.

This PR upgrades the terra classic contracts to the new CosmWasm stdlib version, along with upgrades to the testing pipeline to support the new version. Some contract changes had to be made to accommodate the new 32 byte addressing scheme, these are documented in #3597.

## Reviewing the PR
The PR is best reviewied commit-by-commit. The upgrade to CW1 necessitated vendoring the `cw20-legacy` library. To keep things tidy, I forked it as a separate repo, the relevant commit is here https://github.com/wormhole-foundation/cw20-legacy/commit/d12724701ffb9920cb1fdf1eee49efdcc20048e5.

### CODE REVIEW

**Comment by tonyjin** [comment]
If we go with option 1, do we risk more future breakages?

**Comment by eliiik** [comment]
Any updates? It's still not fixed half a year since then.

File: terra/Makefile [hunk 1]
```diff
@@ -24,6 +24,7 @@ endif
 
 $(WASMS) artifacts/checksums.txt: $(SOURCE_FILES)
 	DOCKER_BUILDKIT=1 docker build --target artifacts -o artifacts .
+	sed -i 's/interface_version_7/interface_version_8/g' $(WASMS)
```
Comment (kcsongor) [inline terra/Makefile:27]: just to remind ourselves, this will not be necessary

Comment (kcsongor) [reply terra/Makefile]: we tried that approach and it was broken in subtle ways, so the current plan is to do option 2. My comment above is meant as a reminder to revert this change, as we're no longer doing it that way

File: devnet/node.yaml [hunk 2]
```diff
@@ -121,7 +121,7 @@ spec:
             # - --terraLCD
             # - http://terra-terrad:1317
             # - --terraContract
-            # - terra18vd8fpwxzck93qlwghaj6arh4p7c5n896xzem5
```
Comment (kcsongor) [inline devnet/node.yaml:124]: there are few copies of this address in the codebase remaining that we should review before merging

File: terra/contracts/token-bridge/src/contract.rs [hunk 3]
```diff
@@ -60,90 +61,9 @@ pub enum TransferType<A> {
...
-    let vaa = parse_vaa(deps.as_ref(), env.block.time.seconds(), &data)?;
-
-    vaa_archive_add(deps.storage, vaa.hash.as_slice())?;
```
Comment (kcsongor) [inline terra/contracts/token-bridge/src/contract.rs:114]: let's double check that this VAA is indeed replay protected on mainnet

File: devnet/node.yaml (hunk 2 above)
Comment (evan-gray) [reply devnet/node.yaml:124]: The two `migrate.js` files are very old (seem like they're there for posterity) and the one in `testing/contract-integrations` isn't used. :shrug:

Comment (kcsongor) [reply devnet/node.yaml:124]: should we just delete them then?

File: terra/contracts/token-bridge/src/contract.rs (hunk 3 above)
Comment (evan-gray) [reply terra/contracts/token-bridge/src/contract.rs:114]: Looks like it is.

I started with

```
worm parse AQAAAAINAJG2fJDyk1xTNYrByMq/IyMPXCSLw7uPBviki4lxrp7VMid+Zg5maFTA7L/hr0wjqg2HvStl1ThaVgs+gtP7r88BA8C6uVKxqz+T35SPMuO5EfRXVppuKk4kgC8AaeMXiT/mKN+G4Ywhq7BhtDfFWBT8r/CONWcr4ukcR4WI9SLo6egABJWNvBNzS87mGaYxNIbh9IJRz8IWt/bUbEM01IHxYDLlcKyF+F5XGcbgtolO9BH0SqJiCSuZf0p2Q3i6JbQPqZQABljhQuTzWBrfh+e/tKwo4woE3I7m7ypwgixDiYu5M7NmE1yRlK/1pvqZDKEQQuwRNeF2bB9Feig95jNhyOl8m4AACOQWt3zZAgFON5iebfkfbGR2j0ct0aYCfZgDLS3Tbs+Zb65tN5otY/JB+276JBsV+neyerpYjEv40aw+myj/9hwACkUWUVQ+fiU3PbQI35LLv1bVEUxMnjNsNnVAx2Zcx3trE/c05ODKWLmcgZC7kYR1TXEUgJDEyR7aRx6lh0fYTNMBC5XgNXIgV0QMepMo+rqEO8GSLje0KgiBuESUwerfb0+gbt3HMiCO+4qnGQLHR9H2e0ENo6i+CVd5ArNbUJOZK7IBDMiWbmpGwiTqPKxRDoRjpcDSCVxBU38JkQhA28mJ/swlFisffm8P9mUjFXm75bWqjxWquhLotlD4wHD8n6WKt8YBDThD1TsiQtGY7PpCV1fXc8JGtnxatZAqBiflM9Byo2hEHpVGHoaB2xC+b5ZPnU7SQb/5yXRwxuNGnB07+8P5rUcBDnu+FszGdOsFKLot3vZj8m6dznXAbxi8VHBjf+DZg0b4VUeAEJO5y4iAHEJf+vzXyaHDuw6uaZy/nE6Cmh2G2WoBDxHFcHxwEmPPImzO3qd/ccUiLE8WLVD7awpYQzuSRhbPbBfluVx7ayBD4tzojqQmO0ST+2FWVjZLYIc2o8a6cwsBEfMoHr6CDGnRx/JYvffw1TCQN4bc736cjPvrto++VciZNXE7/EEWyM8oFiDUDX58COwoZacwkE9ujwlPX9MutOEBEqVTBk6s5r136OH6OXNi7SmCJiMwiN+cNghtIKvXt+dvONA56YZrzsTu2N0skC9xgykjGyxGQBmKTI18I/3r9cMAAAAAAORtswQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEXKpMi7kC0wYgAAAAAAAAAAAAAAAAAAAAAAAAAAAAVG9rZW5CcmlkZ2UCAAMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAXHw==
```

which results in a digest of

```
0xaf6a8f1929882dd02a5f8412be273c36d9ce9818d0181f475abbe63448718be7
```

If we drop into `node` and concatenate that with the key used by `vaa_archive_add`

```
> Buffer.from("\x00\fguardian_set",'ascii').toString('hex') + "af6a8f1929882dd02a5f8412be273c36d9ce9818d0181f475abbe63448718be7".toUpperCase()
'000c677561726469616e5f736574AF6A8F1929882DD02A5F8412BE273C36D9CE9818D0181F475ABBE63448718BE7'
```

We can then get the that from the public endpoint by paginating on the key just before it.

```
> Buffer.from("000C677561726469616E5F736574AF6A8F1929882DD02A5F8412BE273C36D9CE9818D0181F475ABBE63448718BE6",'hex').toString('base64')
'AAxndWFyZGlhbl9zZXSvao8ZKYgt0CpfhBK+Jzw22c6YGNAYH0dau+Y0SHGL5g=='
```

https://terra-classic-lcd.publicnode.com/cosmwasm/wasm/v1/contract/terra10nmmwe8r3g99a9newtqa7a75xfgs2e8z87r2sf/state?pagination.key=AAxndWFyZGlhbl9zZXSvao8ZKYgt0CpfhBK%2BJzw22c6YGNAYH0dau%2BY0SHGL5g%3D%3D&pagination.limit=1

```
{
  "models": [
    {
      "key": "000C677561726469616E5F736574AF6A8F1929882DD02A5F8412BE273C36D9CE9818D0181F475ABBE63448718BE7",
      "value": "dHJ1ZQ=="
    }
  ],
  "pagination": {
    "next_key": "AAxndWFyZGlhbl9zZXSvauH1bEySKG8sKxiUPHgmTHgHDM93NOkOkKdjxrVIRQ==",
    "total": "0"
  }
}
```

and again `node` can confirm that the value is `true`

```
> Buffer.from("dHJ1ZQ==",'base64').toString('ascii')
'true'
```

File: devnet/node.yaml (hunk 2 above)
Comment (evan-gray) [reply devnet/node.yaml:124]: Maybe, but I added the change to https://github.com/wormhole-foundation/wormhole/pull/3594 for now. Didn't want to hold up review of this.

**Review by evan-gray — APPROVED** [review]

File: terra/devnet/config/genesis.json [hunk 4]
```diff
@@ -431,7 +429,7 @@
     },
     "distribution": {
       "params": {
-        "community_tax": "0.000000000000000000",
+        "community_tax": "0.020000000000000000",
```
Comment (eliiik) [inline terra/devnet/config/genesis.json:432]: The community tax is 0.5% now
https://station.terra.money/proposal/columbus-5/11515

Comment (kcsongor) [reply terra/devnet/config/genesis.json:432]: Thanks @eliiik. Personally I think the exact tax rate we test against doesn't matter so much, what's important is that we have a non-0 value here to ensure that the tax handling logic works as intended. The actual number will likely keep changing anyway.

**Review by bruce-riley — DISMISSED** [review]
I only reviewed the go code.

File: docs/devnet.md [hunk 5]
```diff
@@ -21,9 +21,9 @@
...
+| Test CW20          |   Terra   |                                      terra1zwv6feuzhy6a9wekh96cd57lsarmqlwxdypdsplw6zhfncqw6ftqynf7kp | Tokens minted to Test Wallet                                                                                                                                        |
+| Bridge Core        |   Terra   |                                      terra14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9ssrc8au |                                                                                                                                                                     |
+| Token Bridge       |   Terra   |                                      terra1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrquka9l6 |                                                                                                                                                                     |
```
Comment (panoel) [inline docs/devnet.md:26]: Terra and Terra2 have the same addresses by design?

File: sdk/js/src/token_bridge/__tests__/terra-integration.ts [hunk 6]
```diff
@@ -48,29 +49,46 @@ import {
...
 
-// Temporarily disable terra tests until LocalTerra can be upgraded to support v2.1.1
-describe.skip("Terra Integration Tests", () => {
+function sleep(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+// look, broadcast and broadcastBlock still resulted in sequence mismatches
+// and nobody has time for that
+async function broadcastAndWait(terra: LCDClient, tx: Tx) {
```
Comment (panoel) [inline sdk/js/src/token_bridge/__tests__/terra-integration.ts:64]: Should there be a return type listed here?

File: docs/devnet.md (hunk 5 above)
Comment (kcsongor) [reply docs/devnet.md:26]: Yeah, the contract address is derived from the `codeId` (incrementally assigned globally) and the `instantiationId` (incrementally assigned per `codeId`). The actual derivation logic is here https://github.com/cosmwasm/wasmd/blob/017f91ccc9affbe030867925ccb11a7ad49133a7/x/wasm/keeper/addresses.go#L36-L41, it involves hashing those plus some constants and padding bytes.

This means that all cosmwasm chains will share the same 32 byte address for these contracts, though the human-readable (Bech32) representation will depend on the first part of the address, the "human readable part". Terra and Terra2 share this part, it's `terra` in both cases, so the human readable address is rendered the same.

**Review by panoel — DISMISSED** [review]

**Comment by kcsongor** [comment]
had a conflict so had to rebase. I'm also going to move out the first commit (removing the migration code) into a separate PR so we can just squash this PR when merging in, as all the other commits fit together neatly.

**Comment by kcsongor** [comment]
for those who'd like to re-review the recent force push, here's the diff https://github.com/wormhole-foundation/wormhole/compare/3567b59af39cf10606b744a57c54013041cf50ff..ckiss/fix-groundbreaking-terra-upgrade

**Review by evan-gray — DISMISSED** [review]

**Review by evan-gray — APPROVED** [review]

File: terra/test/src/__tests__/bridge.ts [hunk 7]
```diff
@@ -112,7 +124,8 @@ describe("Bridge Tests", () => {
           WASM_MOCK_BRIDGE_INTEGRATION,
           {
             token_bridge_contract: tokenBridge,
-          }
+          },
+          "mockBrigeIntegration"
```
Comment (a5-pickle) [inline terra/test/src/__tests__/bridge.ts:128]: Typo

File: terra/test/src/__tests__/bridge.ts [hunk 8]
```diff
@@ -529,15 +549,17 @@ describe("Bridge Tests", () => {
           mockBridgeIntegration,
           denom
         );
-        const contractExpectedChange = new Int(amount);
+        // tax applied, so we expect less than the original amount
+        const contractBlanceAfterExpected = new Int("99900099");
```
Comment (a5-pickle) [inline terra/test/src/__tests__/bridge.ts:553]: Typo

**Review by a5-pickle — APPROVED** [review]
Reviewed terra sub-directory. Nicely done

File: terra/test/src/__tests__/bridge.ts (hunk 7 above)
Comment (a5-pickle) [reply terra/test/src/__tests__/bridge.ts:128]: ^ just a note. Resolving the conversation.

File: terra/test/src/__tests__/bridge.ts (hunk 8 above)
Comment (a5-pickle) [reply terra/test/src/__tests__/bridge.ts:553]: ^ just a note. Resolving the conversation.

**Pull request merged**